Accessibility: Add header tags missing in title sections

marianunez opened the pull request:
#### Summary

Screen readers navigate through screens better when titles have the proper heading markup. We had several areas that had headers without the proper heading tag, all secondary so changes were to `<h2>` . Headers change include: 

- Create Account title
- Modal headers 
- Autocomplete popup (SuggestionList)
- Search popup: Hints title and suggestion list header
- RHS: Channel members, Search results and Channel info title. 


#### Ticket Link
<!--
If applicable, please include both or either of the following links:

Fixes https://github.com/mattermost/mattermost/issues/XXX
Jira https://mattermost.atlassian.net/browse/MM-XXX
-->
https://mattermost.atlassian.net/browse/MM-63043
https://mattermost.atlassian.net/browse/MM-63019
https://mattermost.atlassian.net/browse/MM-63002
https://mattermost.atlassian.net/browse/MM-61669
https://mattermost.atlassian.net/browse/MM-61621
https://mattermost.atlassian.net/browse/MM-61606
https://mattermost.atlassian.net/browse/MM-61585
https://mattermost.atlassian.net/browse/MM-61584

#### Screenshots
<!--
If the PR includes UI changes, include screenshots/GIFs.

For an easier comparison of UI changes a table (template below) can be used.

|  before  |  after  |
|----|----|
| <insert before screenshot here> | <insert after screenshot here> |

-->

#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates).
* API additions—new endpoint, new response fields, or newly accepted request parameters.
* Database changes (any).
* Schema migration changes. Use the [Schema Migration Template](https://docs.google.com/document/d/18lD7N32oyMtYjFrJKwsNv8yn6Fe5QtF-eMm8nn0O8tk/edit?usp=sharing) as a starting point to capture these details as release notes. 
* Websocket additions or changes.
* Anything noteworthy to a Mattermost instance administrator (err on the side of over-communicating).
* New features and improvements, including behavioral changes, UI changes, and CLI changes.
* Bug fixes and fixes of previous known issues.
* Deprecation warnings, breaking changes, or compatibility notes.

If no release notes are required, write NONE. Use past-tense. Newlines are stripped.

Examples:

```
Added new API endpoints POST /api/v4/foo, GET api/v4/foo, and GET api/v4/foo/:foo_id.
```

```
Added a new config setting ServiceSettings.FooBar. Added a new column Foo to the Users table.
```

```release-note
NONE
```
-->
```release-note
None
```

marianunez commented:
> E2E test has completed for commit `843106054904f26590c683c5ea8f8784446650b7`. Results summary:
> 
> * Cypress: pass rate is `100.00%` (see [Automation Dashboard](https://automation-dashboard.vercel.app/cycle/14495103878_1-8431060-pr-onprem-ent) and commit status check `E2E Tests/test`).
> * Playwright: pass rate is `98.65%` (see commit status check `E2E Tests/test-playwright`).

There is like a pixel difference in the signup screen that is causing the test to fail. 

@matthewbirtch can you double check this screen and see if the margins of `Create your account` is good or requires tweaking?

https://mattermost-pr-30776-4k39y.test.mattermost.cloud/signup_user_complete

<img width="1465" alt="Screenshot 2025-04-16 at 11 45 52 AM" src="https://github.com/user-attachments/assets/6dfabad8-bc52-4eeb-a90e-1c1aeef452fc" />

matthewbirtch commented:
> There is like a pixel difference in the signup screen that is causing the test to fail.
> 
> @matthewbirtch can you double check this screen and see if the margins of `Create your account` is good or requires tweaking?

@marianunez I think it looks fine. No need to tweak in my opinion. However, the space above 'create account' button seems like it needs to be collapsed. There is a margin-top on the button and a margin-bottom on the checkbox input, so we can remove one of those. I assume that's not related to this PR, but would be good to fix. Happy to do that in a follow-up if you'd rather.
<img width="630" alt="Screenshot 2025-04-16 at 12 55 56 PM" src="https://github.com/user-attachments/assets/cacca7db-7592-42b7-8821-6e15c44ad5ea" />

matthewbirtch commented:
@marianunez do you want to me to review more beyond this specific piece?

marianunez commented:
> @marianunez do you want to me to review more beyond this specific piece?

@matthewbirtch If you can do a quick pass on the other areas that the title header were changed but hopefully no styling was affected: 

- Modal headers
- Autocomplete popup (SuggestionList)
- Search popup: Hints title and suggestion list header
- RHS: Channel members, Search results and Channel info title.

marianunez commented:
> There is a margin-top on the button and a margin-bottom on the checkbox input, so we can remove one of those. I assume that's not related to this PR, but would be good to fix. Happy to do that in a follow-up if you'd rather.

Fixed:
<img width="1444" alt="Screenshot 2025-04-16 at 1 54 18 PM" src="https://github.com/user-attachments/assets/6d980833-53b6-4067-b018-1ff403e60d59" />

matthewbirtch reviewed (changes requested):
Hey @marianunez A few things I noticed in my review:

**Create your account page**
- The text for the checkbox "I would like to receive..." is not using the `<label>` tag and therefore cannot be clicked to check off the box. We should change the text element to a `<label>` tag instead of a `<span>` tag and add the `for=""` attribute to the label that links it to the id of the checkbox input. Maybe this is covered in a different accessibility ticket?
- I also noticed the color is different for the first line of this text. I don't think that's necessary.

**Search**
- The `USERS` header in the from: autocomplete is using the wrong font. It should be using the body font `Open Sans` instead of `Metropolis`

**Modals**
- I see you made a change to the modal header component to use `<h2>` instead of `<h1>`, but that doesn't seem to be applying to all modals. Is that a different component for other modals?

**RHS**
- [MM-61621](https://mattermost.atlassian.net/browse/MM-61621) specifically calls out "Info | Town Square" as an example that should use the `<h2>` tag, but only "Info" is wrapped in the tag. The channel name subtitle is outside of the `<h2>`. I see the same thing in the members RHS. 
    <img width="920" alt="image" src="https://github.com/user-attachments/assets/c3472e58-5224-40fc-b193-a68ff0d48b52" />
- For the Recent Mentions, Saved, and Search results RHS views, the titles in the header are sitting too low now. The team name, channel name and RHS title should all be aligned on the same baseline. 
    <img width="1129" alt="Screenshot 2025-04-16 at 2 51 44 PM" src="https://github.com/user-attachments/assets/6df09075-0ec1-46bb-a370-9f452a5510d3" />
- Also, seeing nested `<h2>` tags being used for this title in the rendered HTML
    <img width="318" alt="image" src="https://github.com/user-attachments/assets/10db8e15-e04b-4190-a8f1-52d6ac38d21c" />

marianunez commented:
> * The text for the checkbox "I would like to receive..." is not using the `<label>` tag and therefore cannot be clicked to check off the box. We should change the text element to a `<label>` tag instead of a `<span>` tag and add the `for=""` attribute to the label that links it to the id of the checkbox input. Maybe this is covered in a different accessibility ticket?
> * I also noticed the color is different for the first line of this text. I don't think that's necessary.

@matthewbirtch this issues are not related to this PR changes so we can open a ticket for these to keep this PR within scope of the headers. 

> I see you made a change to the modal header component to use <h2> instead of <h1>, but that doesn't seem to be applying to all modals. Is that a different component for other modals?

Yes, unfortunately not all modals are using the Generic Modal and this one was to cover the LevelAccess report https://mattermost.atlassian.net/browse/MM-61669. Which other modals did you see with the problem and I can see if I can add the change there too? 

> Search and RHS

I'll take a look to fix those styling issues

pvev commented on webapp/channels/src/components/channel_info_rhs/header.tsx:
```diff
@@ -20,8 +20,11 @@ const Icon = styled.i`
 `;
 
-const HeaderTitle = styled.span`
+const HeaderTitle = styled.h2`
```
nit: the plan is start removing styled components from the code base , if possible when dealing with one we should just end its misery.

pvev commented on webapp/channels/src/components/channel_members_rhs/header.tsx:
```diff
@@ -17,8 +17,11 @@ interface Props {
 }
 
-const HeaderTitle = styled.span`
+const HeaderTitle = styled.h2`
```
same here, but feel free to totally ignore it this time, just bringing awareness about it.

pvev reviewed (approved):
🎉

marianunez commented:
@matthewbirtch 

> **Search**
> 
> * The `USERS` header in the from: autocomplete is using the wrong font. It should be using the body font `Open Sans` instead of `Metropolis`

Fixed

> **RHS** [MM-61621](https://mattermost.atlassian.net/browse/MM-61621) specifically calls out "Info | Town Square" as an example that should use the `<h2>` tag, but only "Info" is wrapped in the tag. The channel name subtitle is outside of the `<h2>`. I see the same thing in the members RHS.
Nice catch! The nested h2 was actually fixed by @pvev in a previous PR and I didn't notice. I reverted my change but added it to the Edit History case that it was indeed missing. 

This change also seemed to fix the alignment issue so take a look and let me know.

marianunez commented:
/update-branch

hmhealey commented on webapp/channels/src/components/channel_info_rhs/header.tsx:
```diff
@@ -41,23 +41,25 @@ const Header = ({channel, isArchived, isMobile, onClose}: Props) => {
-                    {channel.display_name}
-                </span>
-                }
+                    {channel.display_name &&
```
I don't think the channel name should be inside of the `h2` tag for this and the `ChannelMembersRhs` below. It arguably could be, but we have the channel name outside of it in the search results at least

hmhealey replied:
Ah, I missed that Matt had suggested that above.

We also discussed this earlier in our 1-1 (before I read the above comment), and agreed on including these in the `h2` elements so that it'd be read out by the screen reader

hmhealey commented on webapp/channels/src/components/new_search/search_hint.tsx:
```diff
@@ -98,10 +104,12 @@ const SearchHints = ({onSelectFilter, searchType, searchTerms, searchTeam, hasSe
-                id='search_hint.filter'
-                defaultMessage='Filter your search with:'
-            />
+            <h2>
```
I know that we were asked to add a heading here, but that feels wrong since these options correspond to the search box above this. 0/5 on if we want to revert this change and argue about that or leave it as-is though

hmhealey replied:
I meant that the new heading moves the buttons into their own section of the page, separate from the search box which those buttons control, and that's what feels wrong to me

matthewbirtch replied:
@hmhealey if not a heading, what element do think makes more sense then?

hmhealey replied:
I'm having a hard time finding anything that matches the visible UX here, and I can't think of any examples of other apps that use a similar pattern at the moment (Slack and Discord both offer those flags as autocomplete results). Looking at just HTML elements, the closest thing I can find is a [`fieldset`](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/fieldset) wrapping the "Filter your search with" text and buttons with the text in a `legend` element, but that doesn't group them with the search box which also seems incorrect.

As it stands, I think it's best to go with Level Access's suggestion for now since I can't find a better way of representing this which matches our visible UX

hmhealey reviewed (approved):
Approving this since I'm still 0/5 on that last thing

yasserfaraazkhan commented:
> @yasserfaraazkhan there is like a pixel difference in the signup screen that is causing the Playwright test to fail. How can I update the image it's using to compare for this test?
> 
> Cypress failure seems irrelevant to my changes.

@marianunez based on [this readMe.](https://github.com/mattermost/mattermost/blob/master/e2e-tests/playwright/README.md?plain=1#L90)

- In terminal 1,  have server and web app running.
- In terminal 2, from the project root, we need to run this docker `docker run -it --rm -v "$(pwd):/mattermost/" --ipc=host mcr.microsoft.com/playwright:v1.52.0-noble /bin/bash`
- Within the docker shell , we can [export these](https://github.com/mattermost/mattermost/blob/master/e2e-tests/playwright/README.md?plain=1#L96:L98)
- Then `cd mattermost/e2e-tests/playwright`
- We can [run this ](https://github.com/mattermost/mattermost/blob/master/e2e-tests/playwright/README.md?plain=1#L113) command to update visual screenshots `npm run test -- visual --update-snapshots`

matthewbirtch commented on webapp/channels/src/components/channel_notifications_modal/__snapshots__/channel_notifications_modal.test.tsx.snap:
```diff
@@ -2713,12 +2713,12 @@ Object {
                           Notification Preferences
-                        </h1>
+                        </h2>
                         <div
```
Given what we did in the RHS for the divider and channel name, I assume we should be consistent here too

marianunez replied:
My only concern with this one is that the structure of the Modal header includes an extra `<div>` for the separator of the title from the subtitle, so adding the header for both would be: 

```
<h2>
    <p> Title </p>
    <div> -- divider
    <p> Subtitle </p>
</h2>
```

Any concerns with this @matthewbirtch @hmhealey ?

hmhealey replied:
I'm good with including the divider and subtitle in the `h2` tag for consistency, but we'd want to change the subtitle to be a `span` instead of a `p` then. From what I can find, it doesn't look like you're supposed to put `p` tags inside of a heading tag

matthewbirtch reviewed (approved):
Approved with just one question
